c10/core/TensorImpl.cpp

```diff
@@ -537,17 +538,24 @@ template <typename VariableVersion>
 c10::intrusive_ptr<TensorImpl> TensorImpl::shallow_copy_and_detach_core(
     VariableVersion&& version_counter,
     bool allow_tensor_metadata_change) const {
-  if (key_set_.has(DispatchKey::Python) &&
+  c10::intrusive_ptr<TensorImpl> r;
+  const auto& maybe_torch_dispatch_mode_state =
+      c10::impl::TorchDispatchModeTLS::get_state();
+  // TODO: do we have to exclude after Python dispatch key set?
```

Comment:
What do you mean by "exclude after Python dispatch key set"?

Reply:
I noticed that in PythonFallback, before calling into Python we exclude all dispatch keys after the Python key. I'm not sure why this is being done (@albanD knows maybe?) but I haven't replicated that logic here.

Reply:
Ho that sounds wrong, it should be excluding all *before* the Python key IIUC

Reply:
here's the relevant code

```
void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
  TORCH_INTERNAL_ASSERT(tls_on_entry.has_value());
  // c10::impl::ForceDispatchKeyGuard dispatcher_guard(tls_on_entry.value());
  // StashTLSOnEntryGuard stash_guard;
  c10::impl::ExcludeDispatchKeyGuard guard(after_Python_keyset);
```